Add support for other entities on the clipboard

MikeFE opened the pull request:
Add clipboard/export support to repos and actors

This new feature adds 'add to clipboard' buttons for both repositories and
actors. The clipboard now has an additional drop down box to allow the user to
change the clipboard view between the three entity types (archival descs,
repositories and actors), to clear the clipboard for each entity type, and to
export each entity type.

Asynchronous jobs to export actors (both CSV and EAC XML), and repositories
(CSV) were also added. Additionally, the export repository task that was
in its own branch was integrated into this work.

sortPicker has been generalized into genericPicker where you can specify the
drop down label and options. You can still call sortPicker as it just invokes
genericPicker with sort-specific labels hard coded.

To summarize:

- Archival descriptions, actors and repositories can now all be added to
  clipboard

- Asynchronous jobs to export all these three entity types are in place

- You can switch between entity types in the clipboard via a dropdown menu

- You can clear the clipboard of each entity type depending on the currently
  selected type.

- Made a more generic sortPicker, "genericPicker."

Commit:
Make 'standards' config file optional for exporters

Conflicts:
	lib/flatfile/QubitFlatfileExport.class.php

Commit:
Handle optional YAML sections more gracefully

Commit:
Initial repository export yaml config/task

Still have yet to export non-directly mapped columns

Commit:
Start adding non-direct columns to export

Commit:
Latest repo export/import code

Export is pretty much done, import still needs a tad of work

Conflicts:
	lib/task/import/csvRepositoryImportTask.class.php

Commit:
Move clipboard files to more generic location

Commit:
Make drop down picker generic, minor css tweak

Commit:
Adding clipboard btn to actors

Commit:
Show actor clipboard properly, fix format dropdown

Commit:
Add skeleton jobs for actor csv/xml export

Also rename io export job filenames for consistency.

Commit:
Initial EAC export async job

Commit:
Initial clear clipboard suppot for repo/actors

Commit:
Initial export actor CSV class / async job

Commit:
Allow per-entity type clipboard clear action

Commit:
Add clipboard btns to repository browse

Commit:
Make export options auto-select proper ent type

Previously, when a user would click export for e.g. repositories in the
clipboard, the export options screen would always default to archival
description. It will now auto-select the proper entity type that the user
selected before clicking 'export'.

Commit:
Ensure export options reflect initial entity type

Commit:
minor tweak

Commit:
Add actor relations csv export support

Commit:
Add async repository csv export job

Commit:
Add aliases file to auth rec export

Commit:
Improve no results message styling

Commit:
Fix repository csv file path code

jraddaoui commented:
Wow, this is quite a pull request @MikeFE!

I guess it will be too much work to split it in various PR at this point, but I think we should. Or, if we don't have time to do that, at least we should improve a lot the description (and the commit message that will be used in the squashed merge) as it doesn't even talk about the asynchronous jobs.

For the next time we work in big features, I'd love to advice that we start the code review process earlier, that way we could have spotted this sooner, reviewed the clipboard part and merged it, to avoid conflicts in the subsequent PRs, etc. Also, I'd appreciate clearer individual commit messages and better descriptions.

About the code, it looks great overall! But I'll take a deeper look tomorrow or on Monday, just in case we can split the PR or improve the description.

MikeFE commented:
Sounds good Radda. Yeah, this PR got way larger than originally anticipated since this feature went way over hours and we vastly under estimated how much work we were going to do on it unfortunately, but I'm totally open to discussion on how we can avoid this sort of thing in the future.

regards

jraddaoui commented:
Great @MikeFE! Please, could you improve the description then? Thanks!

MikeFE commented:
OK I rewrote the PR description and gave it my best shot. Hopefully that's what you were looking for :)

jraddaoui commented on apps/qubit/modules/default/templates/_genericPicker.php:
```diff
@@ -0,0 +1,39 @@
+
```
We don't use to start a template with an empty line

jraddaoui commented on apps/qubit/modules/default/templates/_sortPicker.php:
```diff
@@ -1,43 +1,2 @@
-
-</div>
+<?php echo get_partial('default/genericPicker', array('options' => $sf_data->getRaw('options'),
+                       'param' => 'sort', 'label' => __('Sort by'))) ?>
```
But we use to add a new line at the end of all files. Check: http://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

If you use Sublime Text, that can be set in a config setting, so it automatically adds a new line at the end if it's not there when saving.

MikeFE replied:
hmm OK I use VSCode and didn't notice it wasn't adding the newlines! I'll look into it :)

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -82,23 +115,24 @@ protected function doBackgroundExport($request)
+      'params' => array('fromClipboard' => true, 'slugs' => $this->context->user->getClipboard()->getAll()),
+      'current-level-only' => 'on' !== $request->getParameter('includeDescendants'),
+      'public' => 'on' !== $request->getParameter('includeDrafts'),
+      'objectType' => $request->getParameter('objectType'),
```
You could use `$this->objectType` in here as it was obtained before with a default value

jraddaoui commented on apps/qubit/modules/repository/templates/_browseCardView.php:
```diff
@@ -21,7 +21,7 @@
       </a>
       <div class="bottom">
-        <p><?php echo $authorizedFormOfName ?></p>
+        <p><?php echo get_component('object', 'clipboardButton', array('slug' => $doc['slug'], 'wide' => true)) ?><?php echo $authorizedFormOfName ?></p>
```
I think we should use the small clipboard button and place it in the right side as it takes a lot of place in those cards footer. You may need to fix the text ellipsis or place the button in another container.

jraddaoui commented on apps/qubit/modules/repository/templates/_browseTableView.php:
```diff
@@ -53,6 +53,8 @@
         <?php echo link_to(render_title(get_search_i18n($doc, 'authorizedFormOfName', array('allowEmpty' => false,
                            'culture' => $selectedCulture))), array('module' => 'repository', 'slug' => $doc['slug'])) ?>
+
+        <?php echo get_component('object', 'clipboardButton', array('slug' => $doc['slug'], 'wide' => true)) ?>
```
Same here, it doesn't look right alongside the auth. form of name as it changes the button position. I'd use the small button in a new column at the end of the table, without column header title, so it takes only the button width. We probably don't have time to add the column toggle (All/None) in the column header, like in the description updates page, but we should check it, just in case it's an easy change.

jraddaoui commented on apps/qubit/modules/user/actions/clipboardAction.class.php:
```diff
@@ -25,25 +25,29 @@
 
-    $slugs = $this->context->user->getClipboard()->getAll();
+    // Get entity type class name
+    $this->entityType = 'Qubit'.ucfirst($request->getGetParameter('type', 'informationObject'));
```
This variable could be created after `$this->type` is set, to avoid getting the parameter from the request twice.

jraddaoui commented on apps/qubit/modules/user/actions/clipboardClearAction.class.php:
```diff
@@ -36,6 +38,7 @@ public function execute($request)
+    $this->typeLabel = $this->getTypeLabel();
 
     // Redirect to clipboard page if the clipboard is empty
     if (count($slugs) == 0)
```
We probably need to check `$slugs[$this->type]` in here, like the changes in `clipboardAction.class.php`

jraddaoui commented on lib/flatfile/config/QubitActor-atom.yml:
```diff
@@ -0,0 +1,21 @@
+  - levelOfDetail
+  - revisionHistory
+  - sources
+  - maintenanceNotes
```
New line at the end.

jraddaoui commented on lib/flatfile/config/QubitActor.yml:
```diff
@@ -0,0 +1,28 @@
+
+note:
+
+cacheTaxonomies:
```
;) the same, new line

jraddaoui commented on lib/task/migrate/migrations/arMigration0137.class.php:
```diff
@@ -47,7 +47,7 @@ public function up($configuration)
         $menu->path = 'user/clipboardClear';
         $menu->sourceCulture = 'en';
-        $menu->label = 'Clear selections';
+        $menu->label = 'Clear all selections';
```
I think this is a 2.3 migration, so instances upgraded from 2.3 to 2.4 won't get the label change. I think we must create a new migration script.

jraddaoui commented on plugins/sfIsaarPlugin/modules/sfIsaarPlugin/templates/indexSuccess.php:
```diff
@@ -35,6 +35,11 @@
+
+        <li class="clipboard">
+          <?php echo get_component('object', 'clipboardButton', array('slug' => $resource->slug, 'wide' => true)) ?>
+        </li>
```
I think this button is missing in the repository index page.

jraddaoui reviewed (changes requested):
Hi @MikeFE,

Thank you for the description update, we should use it as the commit body when we merge this work. I'd add a small note about the generic picker changes in there too.

I've added a few comments, some that I think must be addressed before merging and others not so important. I also have the feeling that we could have covered all export cases with only one export job, as there is a lot of duplicated code in those classes, but it may be too late to change it now.

Great work!

Commit:
Code review fixes

- Add newlines at end of files and removed an extra newline at the start of one.

- Use $this->objectType instead since it's already defined & has default val.

- I accidentally removed the include_once for QubitHelper earlier, it was causing
  an error about missing check_element_visibility. Re-added.

- captureResourceExportTemplateOutput made $options a required parameter, this
  isn't in line with the rest of our code and was causing a warning when the
  function was called without specifying options. Made it optional.

- Move variable to avoid duplicate function call

- Make clear clipboard more tightly bound to entity type

Commit:
Add migration for menu label change

Commit:
Change clipboard btn layout for repos

Commit:
Add clipboard btn to repository indexSuccess

Commit:
Restrict sort options based on entity type

Previously, we had only one list of sort options for all entity types in the
clipboard. However, this would cause an error if you selected a sort field that
the newly selected entity type didn't support, e.g. start date for repositories

The sort by picker will now intelligently add only the appropriate sort fields
available given the current entity type.

Commit:
Only filter drafts on information objects

Previously, since the clipboard was information objects only, we were filtering
drafts out in case users added them manually. This commit prevents filtering
drafts for entity types other than info object (which was breaking the results)

Commit:
Restore clear clipboard btn for anonymous users

An isAuthenticated check was preventing non-authenticated users from seeing the
clear clipboard button in the actions toolbar. This rectifies that.

Commit:
Improve entity type dropdown usability for export

Previously, when objectType was set in the URL the drop down for entity type at
the export screen would auto-select the entity type the user was browsing in
the clipboard before hitting the 'export' button. However, the user could then
manually select a different entity type from the dropdown even if that entity
type had 0 clipboard items.

This commit makes it so if entityType is set, only populate the dropdown with
said entityType, not the other ones.

jraddaoui commented on plugins/arDominionPlugin/css/less/buttons.less:
```diff
@@ -170,6 +170,10 @@ button.clipboard-wide {
   }
 }
 
+button.clipboard.clipboard-repository-browse {
```
You could have place this inside the `button.clipboard` section bellow using `&.clipboard-repository-browse`. The transformation to CSS will be the same but it makes a nicer LESS with the nested rules.

jraddaoui commented on lib/task/migrate/migrations/arMigration0150.class.php:
```diff
@@ -0,0 +1,48 @@
+  {
+    if (null !== $menuItem = QubitMenu::getByName('clearClipboard'))
+    {
+      $menuItem->label = 'Clear all selections';
```
Could you remove the changes in lib/task/migrate/migrations/arMigration0137.class.php, please.

sevein replied:
I think we shouldn't change previous migrations, right? We should only update state from newer migrations.

sevein replied:
@MikeFE I think you should replace `$menuItem->label = 'Clear all selections';` with:

```php
$menuItem->setLabel('Clear all selections', array('culture' => 'en'));
```

sevein commented on lib/task/migrate/migrations/arMigration0150.class.php:
```diff
@@ -0,0 +1,48 @@
+    if (null !== $menuItem = QubitMenu::getByName('clearClipboard'))
+    {
+      $menuItem->label = 'Clear all selections';
+      $menuItem->sourceCulture = 'en';
```
This one you don't need it because it should already have that value in it.

jraddaoui commented:
Hi @MikeFE, looking great!

Now that we have the button in both repository view options I think we should add it to the card (digital object) view in the IO browse page, it should be an easy addition. Also, for those card views, since we are placing the clipboard button to the right in all cases, I'd place it in the right bottom corner of the card instead of at the left-bottom corner.

And please, add a little note about the generic picker in the description. Thanks!

jraddaoui commented:
For following PRs, could you avoid to force push until just before the merge. It's not a big deal but it keeps the commit history and review messages in place ;)

Commit:
Improve menu item migration

Use setLabel and set culture instead of sourceCulture.

Commit:
Align clip icon right, add to digiobj browse

Align the clipboard icon to the right when viewing card view of repository.
Also add the clipboard icon to when we're viewing digital objects.

jraddaoui commented:
Hi @MikeFE,

After all, for the clipboard button inside the cards (DO browse and repository browse in card view) we could have avoided passing the parameter and adding the class by adding the CSS rules for `.masonry-brick button.clipboard`, it will remove some code but feel free to leave it be. But please, remove the unneeded changes in lib/task/migrate/migrations/arMigration0137.class.php.

Thanks!

MikeFE commented:
merged into qa/2.4.x 4403962532fb87df6623c335a7fc66f5fbd8b214

jraddaoui commented:
Hi @MikeFE, 

Please, delete the development branches after merge. We're "starting" to have a big list of branches :)

Thanks!!

jraddaoui commented:
Also, the commit message used in the merge doesn't have a reference to the Redmine ticket ...
